clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -632,9 +644,68 @@ public void beginTransaction() throws ProducerFencedException {
      */
     public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
                                          String consumerGroupId) throws ProducerFencedException {
+        if (!cachedGroupMetadata.groupId().equals(consumerGroupId)) {
```

Comment:
The purpose of this `cachedGroupMetadata` is to avoid the creation of groupMetadata everytime we call the old API. It could also be served as a security check on whether the consumer group id has changed in the middle by sending out a warning indicating some illegal state. In Streams or other general EOS use cases, this should never happen.

Reply:
While this is allowed, I couldn't imagine the case as straightforward when you are committing offsets for 2 consumer groups within the same txn. But I agree we probably don't need to log a warning or anything here.

Reply:
Sounds good, we shall abandon the cache and just initialize a new one everytime: it shouldn't be a big memory overhead after I think about it. (maybe 100B every commit interval)